python_modules/dagster/dagster/_check/builder.py

```diff
@@ -396,3 +397,43 @@ def build_check_call_str(
                 return f'{name} if isinstance({name}, {it}) else check.inst_param({name}, "{name}", {it})'
 
         failed(f"Unhandled {ttype}")
+
+
+def build_args_and_assignment_strs(
+    fn_args: Iterable[str],
+    defaults: Mapping[str, Any],
+    kw_only: bool,
+    private_default_empty_container: bool,
+) -> tuple[str, str]:
+    """Utility function to create the arguments to the function as well as any
+    assignment calls that need to happen for default values.
+    """
+    args = []
+    set_calls = []
+    for arg in fn_args:
+        if arg in defaults:
+            default = defaults[arg]
+            if default is None:
+                args.append(f"{arg} = None")
+            # dont share class instance of default empty containers
+            elif private_default_empty_container and default == []:
+                args.append(f"{arg} = None")
+                set_calls.append(f"{arg} = {arg} if {arg} is not None else []")
+            elif private_default_empty_container and default == {}:
+                args.append(f"{arg} = None")
+                set_calls.append(f"{arg} = {arg} if {arg} is not None else {'{}'}")
+            # fallback to direct reference if unknown
+            else:
+                args.append(f"{arg} = {INJECTED_DEFAULT_VALS_LOCAL_VAR}['{arg}']")
+        else:
+            args.append(arg)
+
+    args_str = ""
+    if args:
+        args_str = f", {'*,' if kw_only else ''} {', '.join(args)}"
+
+    set_calls_str = ""
+    if set_calls:
+        set_calls_str = "\n    ".join(set_calls)
+
+    return args_str, set_calls_str
```

Comment:
what does the code generated by this function look like? Maybe document via a test case?

Reply:
this is just moved code, updated test case below [1] shows what it looks like

Reply:
you are consistent
https://github.com/dagster-io/dagster/pull/22374#discussion_r1632241560

Reply:
smart guy!